File: rust/examples/linux/comms/rfm9x/minimal/src/main.rs
use embedded_hal_bus::spi::ExclusiveDevice;
use linux_embedded_hal::SpidevBus;
use periph::chips::comms::Rfm95Minimal;
use spidev::{SpiModeFlags, Spidev, SpidevOptions};

struct NullCs;
impl embedded_hal::digital::ErrorType for NullCs {
    type Error = core::convert::Infallible;
}
impl embedded_hal::digital::OutputPin for NullCs {
    fn set_low(&mut self) -> Result<(), Self::Error> { Ok(()) }
    fn set_high(&mut self) -> Result<(), Self::Error> { Ok(()) }
}

fn main() {
    let dev_path = std::env::var("SPI_DEV").unwrap_or_else(|_| "/dev/spidev0.0".to_string());
    let mut spi = Spidev::open(dev_path).expect("open spi");
    spi.configure(&SpidevOptions::new()
        .max_speed_hz(5_000_000)
        .mode(SpiModeFlags::SPI_MODE_0)
        .build()).expect("configure spi");
    let bus = SpidevBus(spi);
    let device = ExclusiveDevice::new_no_delay(bus, NullCs).expect("spi device");

    let mut radio = Rfm95Minimal::new(device, 868_000_000).expect("init RFM95");    // Create RFM95W driver, (spi, frequency_hz=868e6) → Result

    radio.send(b"hello").expect("send");                                              // Send packet, (data=&[u8] ≤255 B) → Result<()>
    println!("sent");
}

File: cpp/examples/linux/comms/RFM9x/minimal/main.cpp
#include <cstdio>
#include <ctime>
#include <unistd.h>
#include "SPIConnectionLinux.h"
#include "RFM9x.h"

#ifndef TEST_SPI_BUS
#define TEST_SPI_BUS 0
#endif
#ifndef TEST_SPI_DEVICE
#define TEST_SPI_DEVICE 0
#endif

int main() {
    SPIConnectionLinux connection(TEST_SPI_BUS, TEST_SPI_DEVICE, 0, 5000000); // Create SPI connection, (bus=0, device=0, mode=0, max_speed_hz=5e6 Hz) → SPIConnectionLinux
    RFM95Minimal radio(connection, 868000000);           // Create RFM95W driver, (connection, frequency_hz=868e6 Hz) → RFM95Minimal

    const uint8_t msg[] = "hello";
    uint8_t buf[255];
    while (true) {
        radio.send(msg, sizeof(msg) - 1);                // Send packet, (data, len ≤ 255) → void
        size_t len = 0;
        bool ok = radio.receive(buf, len, 2000);         // Receive single packet, (buf, len, timeout_ms=2000 ms) → bool
        if (ok) printf("rx %u B\n", (unsigned)len);
        else    printf("rx timeout\n");
        usleep(3000 * 1000);
    }
}
